File: python/src/x402_seller_starter/find_escrow_payto.py
"""Find sla-escrow payTo via facilitator discovery rail."""

from __future__ import annotations

import json
import os
import sys
import urllib.request


def main() -> None:
    base = os.environ.get("FACILITATOR_BASE_URL", "").rstrip("/")
    wallet = os.environ.get("MERCHANT_WALLET") or os.environ.get("SELLER_WALLET")
    asset = os.environ.get("X402_ASSET", "USDC")
    if not base or not wallet:
        print("Set FACILITATOR_BASE_URL and MERCHANT_WALLET", file=sys.stderr)
        sys.exit(1)
    url = f"{base}/api/v1/facilitator/sellers/{wallet}/rails/sla-escrow?asset={asset}"
    with urllib.request.urlopen(url) as resp:
        info = json.load(resp)
    pay_to = info.get("payTo")
    if not pay_to:
        print("discovery missing payTo", file=sys.stderr)
        sys.exit(1)
    print(f"X402_SCHEME=sla-escrow")
    print(f"X402_PAY_TO={pay_to}")
    print(f"X402_MERCHANT_WALLET={wallet}")


if __name__ == "__main__":
    main()

File: rust/examples/find_escrow_payto.rs
//! Find sla-escrow `payTo` (escrow PDA) via facilitator discovery rail.
//!
//! ```bash
//! MERCHANT_WALLET=<pubkey> FACILITATOR_BASE_URL=https://preview.ipay.sh cargo run --example find_escrow_payto
//! ```

use serde_json::Value;

#[tokio::main]
async fn main() {
    if let Err(e) = run().await {
        eprintln!("Error: {e}");
        std::process::exit(1);
    }
}

async fn run() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let _ = dotenvy::from_path(std::path::Path::new(env!("CARGO_MANIFEST_DIR")).join(".env"));
    let base = std::env::var("FACILITATOR_BASE_URL")?
        .trim_end_matches('/')
        .to_string();
    let wallet = std::env::var("MERCHANT_WALLET").or_else(|_| std::env::var("SELLER_WALLET"))?;
    let asset = std::env::var("X402_ASSET").unwrap_or_else(|_| "USDC".to_string());
    let url = format!("{base}/api/v1/facilitator/sellers/{wallet}/rails/sla-escrow?asset={asset}");
    let info: Value = reqwest::Client::new()
        .get(&url)
        .send()
        .await?
        .error_for_status()?
        .json()
        .await?;

    let pay_to = info["payTo"]
        .as_str()
        .ok_or("discovery response missing payTo")?;

    println!("\n═══ find_escrow_payto (wire-only sla-escrow starter) ═══\n");
    println!("X402_SCHEME=sla-escrow");
    println!("X402_PAY_TO={pay_to}");
    println!("X402_MERCHANT_WALLET={wallet}");
    println!("\nSet ORACLE_AUTHORITIES, ORACLE_PROFILE_ID, ORACLE_NORMATIVE_SPEC_URL in .env.");
    println!("Delivery / SubmitDelivery are out of scope — fund-only wire path.\n");
    Ok(())
}
